### PR TITLE
Use String#-@ in key places to reduce string duplications

### PR DESCRIPTION
**Pull request by casperisfine**
While memory profiling our application, I noticed a lot of duplicated strings being help by graphql structs:

```
Retained String Report
-----------------------------------
1600  "userErrors"
 391  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:120
 391  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/has_fields.rb:45
 391  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/traversal.rb:206
 390  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/object.rb:123
  25  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/language/lexer.rb:1386
   9  /tmp/bundle/ruby/2.5.0/gems/graphql-client-0.14.0/lib/graphql/client/schema/object_type.rb:57
   1  /tmp/bundle/ruby/2.5.0/bundler/gems/json-b5f423ccf08f/lib/json/common.rb:155
   1  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/loader.rb:116
   1  /tmp/bundle/ruby/2.5.0/gems/graphql-client-0.14.0/lib/graphql/client/schema.rb:118

 1571  "name"
 1253  [UNRELATED]
  125  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:164
   81  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/field.rb:177
   31  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/argument.rb:35
   24  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/language/lexer.rb:1386

 1406  "code"
 1336  [UNRELATED]
   26  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:164
   20  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/field.rb:177
    6  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/argument.rb:35

  901  "id"
  310  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:164
  238  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/argument.rb:35
  156  [REDACTED]
   62  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/field.rb:177
   41  [UNRELATED]
   38  [REDACTED]
   27  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/language/lexer.rb:1386

 496  "first"
 257  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:164
 236  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/argument.rb:35
   1  /tmp/bundle/ruby/2.5.0/gems/bootsnap-1.4.3/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22
   1  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/language/lexer.rb:1386

 437  "after"
 225  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:164
 204  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/argument.rb:35

 429  "before"
 221  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/member/build_type.rb:164
 200  /tmp/bundle/ruby/2.5.0/gems/graphql-1.9.4/lib/graphql/schema/argument.rb:35

```

And the list goes on.

After applying this patch the duplications are gone and the app memory usage is reduced by ~1.5MB.

### CODE REVIEW

**Comment by rmosolgo** [comment]
Awesome! Thanks for sharing this fix. I've looked a lot at _runtime_ allocations but not _boot_ allocations. I never thought about how `-` could reduce the retained memory like this. 

_/me tucks this trick in his back pocket_

**Comment by casperisfine** [comment]
Welcome. If you are interested to look into this, here are the memory retention hotspots we're seeing in our app:

```
retained memory by gem
-----------------------------------
  33.95 MB  graphql-ruby-37a29c98bde0

retained memory by location
-----------------------------------
   9.75 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/define/instance_definable.rb:202
   4.54 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/query/arguments.rb:13
   2.68 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/define/instance_definable.rb:107
   1.53 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/schema/field.rb:105
   1.23 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/define/instance_definable.rb:129
    1.2 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/schema/field.rb:332
   1.13 MB  /tmp/bundle/ruby/2.5.0/bundler/gems/graphql-ruby-37a29c98bde0/lib/graphql/define/instance_definable.rb:136
```

I'll likely try to reduce that in the near future, but that isn't very high priority.

**Comment by rmosolgo** [comment]
I'm hoping to remove instance_definable from the normal flow in 1.10, so hopefully that will help! For reference: 

`singleton_class`es: 

https://github.com/rmosolgo/graphql-ruby/blob/168867b7421f1b1a287b0b77a6f80c3e05b49ea8/lib/graphql/define/instance_definable.rb#L202

`metadata` hashes: 

https://github.com/rmosolgo/graphql-ruby/blob/168867b7421f1b1a287b0b77a6f80c3e05b49ea8/lib/graphql/define/instance_definable.rb#L107

https://github.com/rmosolgo/graphql-ruby/blob/168867b7421f1b1a287b0b77a6f80c3e05b49ea8/lib/graphql/define/instance_definable.rb#L136

`.dup`s: 

https://github.com/rmosolgo/graphql-ruby/blob/168867b7421f1b1a287b0b77a6f80c3e05b49ea8/lib/graphql/define/instance_definable.rb#L129

Basically, `.define` and `.redefine` are accumulating cruft 😩  I saw that you made some changes to the lexer, are you using `Schema.from_definition` here? It might be a bit better in 1.10 because I migrated  to use classes instead of `.define`.

**Comment by rmosolgo** [comment]
Oh, and the others: 
I'm not sure if this will be necessary in a `.define`-less world or not: 

https://github.com/rmosolgo/graphql-ruby/blob/61a0669e7e92a9a5a5ec6b1d18d7fbf104f8d3b4/lib/graphql/query/arguments.rb#L13

This is making `GraphQL::Schema::Field` instances, unlikely to ever be removed: 

https://github.com/rmosolgo/graphql-ruby/blob/168867b7421f1b1a287b0b77a6f80c3e05b49ea8/lib/graphql/schema/field.rb#L105

In the `#to_graphql` path which I hope to remove by default in 1.10: 
https://github.com/rmosolgo/graphql-ruby/blob/168867b7421f1b1a287b0b77a6f80c3e05b49ea8/lib/graphql/schema/field.rb#L332

**Comment by casperisfine** [comment]
Yeah I identified that `.define` hotspot, looked at at the redefinition code, and suddenly decided it was late and that I was better to call it a day 😛  

> I saw that you made some changes to the lexer, are you using `Schema.from_definition` here?

Did I? I'm basically science-dog here. I'm not working with GraphQL at all, and only have very basic understanding of the protocol. I'm trying to reduce memory usage across the board and this popped up.

**Comment by rmosolgo** [comment]
> basically science-dog

Haha, fair enough. Hopefully some of these will be "just gone" in 1.10 (or able to be bypassed by using GraphQL::Execution::Interpreter).

**Comment by eapache** [comment]
We're not using `from_definition` 🤷‍♂️ . We're also not using the interpreter, yet (🔜). Once we get that out Scott or I can sit down with Jean and dig into some of the other retentions that are left.

**Comment by casperisfine** [comment]
> We're not using `from_definition` 🤷‍♂️

@eapache Turns out we do 😄 (cusco client).

**Comment by eapache** [comment]
Oooooh, I wasn't thinking about the client at all... right.

**Comment by segunadeleye** [comment]
Unfortunately `String#-@` is available from Ruby 2.3 upwards. It's breaking in Ruby 2.2 😭

**Comment by rmosolgo** [comment]
Oops 😖 Opened a PR to fix at #2305

**Comment by rmosolgo** [comment]
🎉  I got Ruby 2.2 running on Travis again and just released 1.9.6. Give that version a try!

**Comment by casperisfine** [comment]
I didn't know 2.2 was still supported. Sorry about that :/

**Comment by rmosolgo** [comment]
Blame the guy who merged the PR 😉 Thanks for the improvement here.
